### ISSUE TITLE
Default SingleSampleMediaSource.treatLoadErrorsAsEndOfStream to true

### ISSUE THREAD

**Issue by GouravSna** [google/ExoPlayer#8430]
- ExoPlayer version number : 2.12.2 and 2.12.1
- Android version : Android 10 (not OS specific)
- Android device : Not device specific

**Steps to reproduce:**

1. Pass the external subtitle url to media item. Can replace this [method](https://github.com/google/ExoPlayer/blob/107fed9b2afc1e1edce95eee4a1fbc952377efd2/demos/main/src/main/java/com/google/android/exoplayer2/demo/IntentUtil.java#L127) in source code with the following method.

I have appended __xyz_ in the vtt url.

```java
private static List<MediaItem.Subtitle> createSubtitlesFromIntent(
      Intent intent, String extrasKeySuffix) {
//    if (!intent.hasExtra(SUBTITLE_URI_EXTRA + extrasKeySuffix)) {
//      return Collections.emptyList();
//    }
    String url = "https://storage.googleapis.com/exoplayer-test-media-1_xyz/webvtt/numeric-lines.vtt";
    MediaItem.Subtitle subtitleMediaItem = new MediaItem.Subtitle(Uri.parse(url), MimeTypes.TEXT_VTT,"en", C.SELECTION_FLAG_DEFAULT, C.ROLE_FLAG_SUBTITLE, "ENGLISH-ZEE");

    return Collections.singletonList(subtitleMediaItem);
  }
```

2. Run any content and playback will fail.

**Logcat**

```java
2021-01-05 19:14:08.615 31453-31570/com.google.android.exoplayer2.demo E/ExoPlayerImplInternal: Playback error
      com.google.android.exoplayer2.ExoPlaybackException: Source error
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:554)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:165)
        at android.os.HandlerThread.run(HandlerThread.java:61)
     Caused by: com.google.android.exoplayer2.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 404
        at com.google.android.exoplayer2.ext.cronet.CronetDataSource.open(CronetDataSource.java:486)
        at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:199)
        at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource(CacheDataSource.java:764)
        at com.google.android.exoplayer2.upstream.cache.CacheDataSource.open(CacheDataSource.java:579)
        at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:84)
        at com.google.android.exoplayer2.source.SingleSampleMediaPeriod$SourceLoadable.load(SingleSampleMediaPeriod.java:428)
        at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:415)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:760)
```

**Some Background**

Prior exoplayer 2.12.1, we were passing `DefaultLoadErrorHandlingPolicy` and here we were listening to Exception but after 2.12.1 this API does not seem to be working. We tried it but we are not getting any callback.

Following class we have added to listen the callback before 2.12.1 and it was fine. It means if the subtitle does not work, we listen to it and it does not hamper the playback.

``` java

public class ExternalTextTrackLoadErrorPolicy extends DefaultLoadErrorHandlingPolicy {

    private static final PKLog log = PKLog.get("ExternalTextTrackLoadError");

    private ExternalTextTrackLoadErrorPolicy.OnTextTrackLoadErrorListener textTrackLoadErrorListener;

    public interface OnTextTrackLoadErrorListener {
        void onTextTrackLoadError(PKError currentError);
    }

    public void setOnTextTrackErrorListener(ExternalTextTrackLoadErrorPolicy.OnTextTrackLoadErrorListener onTextTrackErrorListener) {
        this.textTrackLoadErrorListener = onTextTrackErrorListener;
    }

    //public void onTextTrackLoadError(PKError currentError) {
    //    textTrackLoadErrorListener.onTextTrackLoadError(currentError);
    //} TODO // need to test with vtt errored file

    @Override
    public long getRetryDelayMsFor(LoadErrorInfo loadErrorInfo) {
        if (loadErrorInfo == null) {
            return Consts.TIME_UNSET;
        }

        IOException exception = loadErrorInfo.exception;
        Uri pathSegment = getPathSegmentUri(exception);
        if (pathSegment == null || !(exception instanceof HttpDataSource.HttpDataSourceException)) {
            return super.getRetryDelayMsFor(loadErrorInfo);
        }


        String lastPathSegment = pathSegment.getLastPathSegment();
        if (lastPathSegment != null && (lastPathSegment.endsWith(PKSubtitleFormat.vtt.pathExt) || lastPathSegment.endsWith(PKSubtitleFormat.srt.pathExt))) {
            PKError currentError = new PKError(PKPlayerErrorType.SOURCE_ERROR, PKError.Severity.Recoverable, "TextTrack is invalid url=" + pathSegment, exception);
            if (textTrackLoadErrorListener != null) {
                log.e("Error-Event sent, type = " + PKPlayerErrorType.SOURCE_ERROR);
                textTrackLoadErrorListener.onTextTrackLoadError(currentError);
            }
            return Consts.TIME_UNSET;
        } else {
            return super.getRetryDelayMsFor(loadErrorInfo);
        }
    }

    private Uri getPathSegmentUri(IOException ioException) {
        DataSpec dataSpec = null;
        if (ioException instanceof HttpDataSource.InvalidResponseCodeException) {
            dataSpec = ((HttpDataSource.InvalidResponseCodeException) ioException).dataSpec;
        } else if (ioException instanceof HttpDataSource.HttpDataSourceException) {
            dataSpec = ((HttpDataSource.HttpDataSourceException) ioException).dataSpec;
        }

        if (dataSpec != null) {
            return dataSpec.uri;
        }
        return null;
    }
}


```

**Comment by icbaker** [comment]
>Prior exoplayer 2.12.1, we were passing DefaultLoadErrorHandlingPolicy and here we were listening to Exception but after 2.12.1 this API does not seem to be working. We tried it but we are not getting any callback.

Can you provide the info about specifically how you tried to use this in 2.12.1?

**Comment by GouravSna** [comment]
It was working in v2.11.7 and we started facing this issue after `MediaItem` introduction in v2.12.x.

In 2.11.7, we were using [SingleSampleMediaSource](https://github.com/kaltura/playkit-android/blob/bc24c6c170fd592df97c6123af6ae26c215c9694/playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java#L443) to set the external subtitles and then apply the error policy on it using  [ExternalTextTrackLoadErrorPolicy.java class](https://github.com/kaltura/playkit-android/blob/dev/playkit/src/main/java/com/kaltura/playkit/player/ExternalTextTrackLoadErrorPolicy.java) and were using [MergingMediaSource](https://github.com/kaltura/playkit-android/blob/bc24c6c170fd592df97c6123af6ae26c215c9694/playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java#L340).

Now with `MediaItem`, this flow is not working and after setting the error policy on `DefaultMediaSourceFactory`; this is not giving us the callback for error.

Please let me know if you want any demo sample code or any more info.

**Comment by icbaker** [comment]
Everything you describe in the original comment is expected until you get to the part about the `LoadErrorHandlingPolicy` registration not working correctly in 2.12.1, so that's what I'd like to dig into.

Please can you provide enough info to reproduce this part of the problem? You haven't made it clear **exactly** how you're registering this policy in using 2.12.1.

Please provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) that demonstrates the problem in a way that we can build locally. This could be an Android Studio project on GitHub, or zipped up and sent to dev.exoplayer@gmail.com using a subject in the format "Issue #1234", where "#1234" should be replaced with your issue number. Please also update this issue to indicate you’ve done this.

**Comment by GouravSna** [comment]
Please check [Sample for Subtitle Not Working](https://github.com/GouravSna/ExoPlayer/tree/PlaybackFailsWithBrokenVTT) branch and [Sample for LoadErrorPolicy](https://github.com/GouravSna/ExoPlayer/tree/LoadErrorHandlingPolicy) branch.

You can see [this is how](https://github.com/GouravSna/ExoPlayer/commit/33765f102cd88dbbb32accf99802d9dec5a2c926) we are adding load error policy.

But we are not getting callback in `getRetryDelayMsFor` method. 
[Here](https://github.com/google/ExoPlayer/issues/6640#issuecomment-553904951) in this comment you gave this solution earlier as well. But now seems that it is broken.

What we can think of that the object which we are giving to exoplayer for error policy is not being retained and internally another object is being created that's why we are not getting the callback in our class.

**Comment by icbaker** [comment]
Thanks for the branch - using that I'm able to reproduce what you're seeing, and using the debugger I pinned down the problem - you're right that basically your policy is being accidentally dropped internally. Specifically here, it needs to be propagated down to the `SingleSampleMediaSource.Factory`:
https://github.com/google/ExoPlayer/blob/107fed9b2afc1e1edce95eee4a1fbc952377efd2/library/core/src/main/java/com/google/android/exoplayer2/source/DefaultMediaSourceFactory.java#L253-L254

I've sent a fix for this - it should be included in 2.12.3 which will be released before the end of the month.

Note that even with this fix, your content doesn't play in your LoadErrorHandlingPolicy branch - but I think that's due to the implementation of your policy - the execution does now reach the `getRetryDelayMsFor` method.

**Comment by GouravSna** [comment]
Hi @icbaker , now the code is reaching to `getRetryDelayMsFor` of my custom class and there I am getting the uri from `dataSpec` and checking if the error is for `vtt` or `srt` and if yes then we are returning `C.TIME_UNSET`.

But there is issue that still the media is failing to play because `treatLoadErrorsAsEndOfStream` is not set to true, it is false and [here](https://github.com/google/ExoPlayer/blob/107fed9b2afc1e1edce95eee4a1fbc952377efd2/library/core/src/main/java/com/google/android/exoplayer2/source/SingleSampleMediaPeriod.java#L296) in this condition, it goes to else part and eventually it retries and fails.

In ExoPlayer v2.11.7, there was an api to pass `treatLoadErrorsAsEndOfStream` value, [Example is this](https://github.com/kaltura/playkit-android/blob/bc24c6c170fd592df97c6123af6ae26c215c9694/playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java#L445) 

I tried setting `treatLoadErrorsAsEndOfStream`  value to `true` in `SingleSampleMediaPeriod` constructor then i can see the media plays back with the load policy.

Can you help to let us know how to pass this value now ?

**Comment by icbaker** [comment]
Ah I see - yeah that isn't currently possible using only the `MediaItem` API.

I will add this as a configuration option to `MediaItem.Subtitle`.

Until that's done you'll need to either implement your own `MediaSource.Factory` or skip using `MediaItem` and use the `MediaSource` APIs:
https://exoplayer.dev/media-sources.html#media-source-based-playlist-api

**Comment by GouravSna** [comment]
Hi @icbaker 
In v2.12.3, only `setLoadErrorHandlingPolicy` is being handled. Please let me know if there is any plan to add `treatLoadErrorsAsEndOfStream`. 
Because I think this fix will not help us.

**Comment by icbaker** [comment]
We plan to handle `treatLoadErrorsAsEndOfStream` in some way (possibly by adding it to `MediaItem.Subtitle`, possibly by defaulting to it to true, or possibly by other means). 

We haven't made this change yet (hence this issue is still open), so yes it's not included in 2.12.3 released today.
